src/types.rs

```diff
@@ -261,3 +262,59 @@ impl BufRingEntry {
         &(*ring_base).0.resv
     }
 }
+
+/// A destination slot for sending fixed resources
+/// (e.g. [`opcode::MsgRingSendFd`](crate::opcode::MsgRingSendFd)).
+#[derive(Debug, Clone, Copy)]
+pub struct DestinationSlot {
+    dest: KernelIndexDest,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum KernelIndexDest {
+    /// Equivalent to `IORING_FILE_INDEX_ALLOC`.
+    Auto,
+    /// Fixed slot as indexed by the kernel (target+1).
+    Index(u32),
+}
+
+impl DestinationSlot {
+    const AUTO_ALLOC: u32 = sys::IORING_FILE_INDEX_ALLOC as u32;
+
+    /// Use an automatically allocated target slot.
+    pub fn auto_target() -> Self {
+        Self {
+            dest: KernelIndexDest::Auto,
+        }
+    }
+
+    /// Try to use a given target slot.
+    ///
+    /// Valid slots are in the range from `0` to `u32::MAX - 2` inclusive.
+    pub fn try_from_slot_target(target: u32) -> io::Result<Self> {
+        const MAX_INDEX: u32 = DestinationSlot::AUTO_ALLOC - 2;
+        match target {
+            v if v <= MAX_INDEX => Ok(Self {
+                dest: KernelIndexDest::Index(v + 1),
+            }),
+            v => Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("Invalid slot target {}", v),
+            )),
+        }
+    }
+
+    pub(crate) fn kernel_index_arg(&self) -> u32 {
+        use KernelIndexDest::*;
+        match self.dest {
+            Index(v) if v == (Self::AUTO_ALLOC - 1) => {
+                unreachable!("IORING_FILE_INDEX_ALLOC - 1")
```

Comment:
This can't actually happen, but writing a `panic!` here won't make the compiler optimize it away.

It would be nice to remove it or use `debug_assert!`.